File: 2017/day4/day4.py
#!/usr/bin/env python3
# [Day 4: High-Entropy Passphrases](https://adventofcode.com/2017/day/4)

import sys
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()

n = 0
for line in lines:
    a = line.split()
    if len(a) == len(set(a)):
        n += 1
print(n)

n = 0
for line in lines:
    a = list(map(lambda x: "".join(sorted(x)), line.split()))
    if len(a) == len(set(a)):
        n += 1
print(n)

File: 2017/day4/day4.rs
//! [Day 4: High-Entropy Passphrases](https://adventofcode.com/2017/day/4)

use std::collections::HashSet;

struct Puzzle {
    lines: Vec<String>,
}

impl Puzzle {
    const fn new() -> Self {
        Self { lines: vec![] }
    }

    /// Get the puzzle input.
    fn configure(&mut self, data: &str) {
        self.lines = data.lines().map(str::to_string).collect();
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        self.lines
            .iter()
            .filter(|line| {
                let a = line.split_ascii_whitespace();
                let n = line.split_ascii_whitespace().count();

                a.collect::<HashSet<_>>().len() == n
            })
            .count()
    }

    /// Solve part two.
    fn part2(&self) -> usize {
        self.lines
            .iter()
            .filter(|line| {
                let a: Vec<_> = line
                    .split_ascii_whitespace()
                    .map(|x| {
                        let mut x: Vec<_> = x.chars().collect();
                        x.sort_unstable();
                        x
                    })
                    .collect();

                a.iter().collect::<HashSet<_>>().len() == a.len()
            })
            .count()
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}
